libglusterfs/src/iobuf.c

```diff
@@ -483,30 +470,18 @@ iobuf_get_from_small(const size_t page_size)
 {
     struct iobuf *iobuf = NULL;
 
-    iobuf = GF_MALLOC(sizeof(*iobuf), gf_common_mt_iobuf);
-    if (!iobuf)
-        goto err;
-
-    iobuf->free_ptr = GF_MALLOC(page_size, gf_common_mt_iobuf_pool);
-    if (!iobuf->free_ptr)
-        goto err;
+    iobuf = GF_MALLOC(sizeof(*iobuf) + page_size, gf_common_mt_iobuf);
```

Comment:
From https://google.github.io/tcmalloc/reference.html :
When __STDCPP_DEFAULT_NEW_ALIGNMENT__ is not specified (or is larger than 8 bytes), we use standard 16 byte alignments for ::operator new without a std::align_val_t argument. However, for allocations under 16 bytes, we may return an object with a lower alignment, as no object with a larger alignment requirement can be allocated in the space. When compiled with __STDCPP_DEFAULT_NEW_ALIGNMENT__ <= 8, we use a set of sizes aligned to 8 bytes for raw storage allocated with ::operator new.

Reply:
I'm surprised GF_IOBUF_ALIGN_SIZE is not 4K btw, I expected it to be.

Reply:
> From https://google.github.io/tcmalloc/reference.html : When **STDCPP_DEFAULT_NEW_ALIGNMENT** is not specified (or is larger than 8 bytes), we use standard 16 byte alignments for ::operator new without a std::align_val_t argument. However, for allocations under 16 bytes, we may return an object with a lower alignment, as no object with a larger alignment requirement can be allocated in the space. When compiled with **STDCPP_DEFAULT_NEW_ALIGNMENT** <= 8, we use a set of sizes aligned to 8 bytes for raw storage allocated with ::operator new.

Yes. I've also checked glibc and other allocators and it seems that for allocations of 16 bytes or more, the alignment will be 16 bytes (on 64-bit systems). So it should be fine for EC at least. For direct I/O I hope we always use `iobuf_get_page_aligned()` to avoid potential errors. In any case this can be clarified and addressed if necessary later.

> I'm surprised GF_IOBUF_ALIGN_SIZE is not 4K btw, I expected it to be.

Yes, I don't know why either. One potential issue here is that glibc (and I would say tcmalloc as well) use mmap for sizes bigger than 128 KiB, which already aligns the offset to a multiple of 4 KiB. That could hide any potential issue caused by a small `GF_IOBUF_ALIGN_SIZE`. With the current change, however, this is broken. The first part of the allocated memory will be used for the struct. So the buffer itself won't be aligned to 4KiB. In this case, if `GF_IOBUF_ALIGN_SIZE` is really insufficient, some problems could appear.